File: src/main/scala/com/vc/hard/L960.scala
package com.vc.hard

object L960 {
  def minDeletionSize(arr: Array[String]): Int = {
    val n = arr.length
    if(n == 0) return 0
    val m = arr(0).length
    val dp = Array.fill(m)(1)
    var res = m - 1

    def checkMaxIncSubSequenceCondForAllRows(i: Int, j: Int): Boolean = {
      var k = 0
      while(k < n) {
        if(arr(k)(i) < arr(k)(j)) return false
        k += 1
      }
      true
    }

    (1 until m).foreach(i => {
      (0 until i).foreach(j => {
        if(checkMaxIncSubSequenceCondForAllRows(i, j))
          dp(i) = Math.max(dp(i), dp(j) + 1)
      })
      res = Math.min(res, m - dp(i))
    })
    res
  }
}

File: src/main/java/com/vc/hard/L960.java
package com.vc.hard;

import java.util.Arrays;

class L960 {
    public int minDeletionSize(String[] arr) {
        //MaxIncreasing Subsequence
        int n = arr.length;
        if(n == 0) return 0;

        int m = arr[0].length();
        int[] dp = new int[m];
        Arrays.fill(dp, 1);
        int res = m - 1;
        for(int i = 1; i < m; i++) {//Check max increasing subsequence at i for all the rows
            for(int j = 0; j < i; j++) {
                if(check(arr, i, j)) {
                    dp[i] = Math.max(dp[i], dp[j] + 1);
                }
            }
            //Remove max increasing subsequence from total length
            res = Math.min(res, m - dp[i]);
        }
        return res;
    }

    private boolean check(String[] arr, int i, int j) {
        for(String str: arr) {
            if(str.charAt(i) < str.charAt(j)) return false;
        }
        return true;
    }
}
